### ISSUE TITLE
Server crashes on connection

### ISSUE THREAD

**Issue by neilio** [shusain93/OSXMessageProxy#6]
I have a feeling this is a caching issue but at any rate I've run into an issue. I had set this up on a new mac that I'm using as a server and it worked as expected the first time I ran it. I quit everything and copied over my Messages folder from my laptop so that the server would have the full archive of messages. I rebooted and the launch services script auto launched the server.

When I try to connect now with the client the server immediately crashes. I've tried rebuilding from scratch and scouring the hard drive for cached data but still can't figure out why it's not working.

This is the pertinent section from the log, but I can post the full thing if that's helpful. Thanks!

```
Crashed Thread:        2  Dispatch queue: com.apple.root.default-qos

Exception Type:        EXC_BAD_INSTRUCTION (SIGILL)
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Illegal instruction: 4
Termination Reason:    Namespace SIGNAL, Code 0x4
Terminating Process:   exc handler [0]

Application Specific Information:
fatal error: unexpectedly found nil while unwrapping an Optional value
```

**Comment by harveson** [comment]
The problem seems to occur when the server's screen is sleeping and requires password to wake. This is because the password box is filled automatically (I guess it is the message).  The crash report is attached.

[crashreport.txt](https://github.com/shusain93/OSXMessageProxy/files/968017/crashreport.txt)

**Comment by harveson** [comment]
I did some experiment. As long as the screen goes to sleep (only the screen not the whole machine), the message will not send. I guess the shortcut is disabled whenever the screen goes to sleep?  I can't keep the server's screen on all the time. It would be great it can support off-screen.

**Comment by shusain93** [comment]
@harveson 
Yeah I use UI automation requires that the server be unlocked and awake the entire time. There's currently no other way to send messages without UI automation. It shouldn't crash, mind you, so thanks for the crash reports.

@neilio 
You are building from source, can you tell paste the line where it crashes when you run from Xcode? It'll crash and Xcode will dump the crash and give you a trace. Click around in the navigator until you get some call in `DatabaseConnector`

**Comment by neilio** [comment]
Thanks for the fast reply, @shusain93. I think this is what you need, at the end?

http://o.neil.io/NSymz

**Comment by shusain93** [comment]
@neilio If you go to ~/Library/Messages do you see a file named chat.db? Does the messaging app work normally? I also copied the database to a new laptop and it works properly. I can't seem to symbolicate the crash so could you also build and run directly from Xcode, no script, and then send a screenshot of the line that it crashes at? Sorry to make you go through this, I've made some changes in the server so I can't get the crash line specifically.

It looks to me like it can't read/find the message database and so it just throws when it tries to read.

**Comment by neilio** [comment]
Yes to the presence of chat.db and Messages. working. Here's a screenshot of the line that gets highlighted.

http://o.neil.io/sjcHs

By the way, when the mobile client connects I see a flash of messages before the server crashes, so it's getting something.

**Comment by shusain93** [comment]
@neilio In that line if you change `as!` to `as?` do you still crash there? You hopefully shouldn't. Also can you make sure that all your conversations show up?

It looks like one of your conversations doesn't have a valid name and when it tries to forcibly read it the application crashes.

**Comment by neilio** [comment]
So the good news is that code change fixed the crashing problem - I now get the conversation list loading properly. The bad news is when I tap into any of the conversations I get a new error on the client that says "Unable to parse json data org.json.JSONexception: Value invalid of type java.lang.String cannot be converted to J..." and then the rest of message is truncated. Anything I can send you to help figure out what's going on?

**Comment by shusain93** [comment]
@neilio If you run the server from the command line you should get a rubbing log of all the data it sends. Could you copy and paste what happens in the server log when you pick a conversation? Does this happen for all of them? Also, did this happen previously? Something is very odd with your message database

**Comment by shusain93** [comment]
Alright so I pulled up the error in the crash reporting software, it hits a parse error immediately when it reads the conversation response. To fix this I'll definitely need to see what data it's returning because I've never seen this before. You've won the broken software lottery, sorry haha.

**Comment by neilio** [comment]
I'm good at breaking things, it's true. I just refreshed here and when I build now I get an error. Ha ha!

http://o.neil.io/yybKC

**Comment by neilio** [comment]
I just tried the pre-compiled binary and it totally works, so I'm wondering if this should be closed as a pebkac.

**Comment by shusain93** [comment]
I think the issue is that I improperly committed the pods. I get this problem on my laptop when building so if you just run `pod deintegrate` and then `pod install` it'll work.

I'll push a new binary later tonight with that other fix you found. It shouldn't hurt anyways even if it's working. It'll bring auto updating which should help with this oddity with being on all sorts of versions at once.

**Comment by neilio** [comment]
Yup, pulling down the pods fixed the build issue. I think this one can be closed.

**Comment by shusain93** [comment]
Let me know if the latest released binary works for you. You should be able to autoinstall it if you're running v1.1 (it'll have the GUI config). It's got a few other fixes and I want to make sure it still works and updates properly for you

**Comment by neilio** [comment]
It seemed to update okay, though the version number in the UI is still 1.1.

**Comment by shusain93** [comment]
It might not be installing then. I assume it goes through the process fine, no errors? When you hit install does it just quit and you have to launch it yourself? I had this issue on one test machines and couldn't find anything in my logs and I hoped it was a localized issue.

**Comment by shusain93** [comment]
I solved the updater issue in version 1.2
